Plugins/FpeMonitoring/src/FpeMonitor.cpp

```diff
@@ -0,0 +1,344 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2022-2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Plugins/FpeMonitoring/FpeMonitor.hpp"
+
+#include "Acts/Utilities/Helpers.hpp"
+
+#include <algorithm>
+#include <bitset>
+#include <cfenv>
+#include <csignal>
+#include <cstddef>
+#include <cstdint>
+#include <iostream>
+#include <iterator>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string_view>
+#include <vector>
+
+#include <boost/stacktrace/frame.hpp>
+#include <boost/stacktrace/safe_dump_to.hpp>
+#include <boost/stacktrace/stacktrace.hpp>
+#include <boost/stacktrace/stacktrace_fwd.hpp>
+#include <fenv.h>
+#include <signal.h>
+
+#define FPU_EXCEPTION_MASK 0x3f
+#define FPU_STATUS_FLAGS 0xff
+#define SSE_STATUS_FLAGS FPU_EXCEPTION_MASK
+#define SSE_EXCEPTION_MASK (FPU_EXCEPTION_MASK << 7)
+
+namespace Acts {
+
+FpeMonitor::Result::FpeInfo::~FpeInfo() = default;
+
+FpeMonitor::Result::FpeInfo::FpeInfo(const FpeInfo &other)
+    : count{other.count},
+      type{other.type},
+      st{std::make_unique<boost::stacktrace::stacktrace>(*other.st)} {}
+
+FpeMonitor::Result::FpeInfo::FpeInfo(
+    std::size_t countIn, FpeType typeIn,
+    std::unique_ptr<boost::stacktrace::stacktrace> stIn)
+    : count{countIn}, type{typeIn}, st{std::move(stIn)} {}
+
+FpeMonitor::Result FpeMonitor::Result::merged(const Result &with) const {
+  Result result{};
+
+  for (unsigned int i = 0; i < m_counts.size(); i++) {
+    result.m_counts[i] = m_counts[i] + with.m_counts[i];
+  }
+
+  std::copy(with.m_stracktraces.begin(), with.m_stracktraces.end(),
+            std::back_inserter(result.m_stracktraces));
+  std::copy(m_stracktraces.begin(), m_stracktraces.end(),
+            std::back_inserter(result.m_stracktraces));
+
+  result.deduplicate();
+
+  return result;
+}
+
+void FpeMonitor::Result::merge(const Result &with) {
+  for (unsigned int i = 0; i < m_counts.size(); i++) {
+    m_counts[i] = m_counts[i] + with.m_counts[i];
+  }
+
+  std::copy(with.m_stracktraces.begin(), with.m_stracktraces.end(),
+            std::back_inserter(m_stracktraces));
+
+  deduplicate();
+}
+
+void FpeMonitor::Result::add(FpeType type, void *stackPtr,
+                             std::size_t bufferSize) {
+  auto st = std::make_unique<boost::stacktrace::stacktrace>(
+      boost::stacktrace::stacktrace::from_dump(stackPtr, bufferSize));
+
+  auto it = std::find_if(
+      m_stracktraces.begin(), m_stracktraces.end(), [&](const FpeInfo &el) {
+        const auto &fl = *el.st->begin();
+        const auto &fr = *st->begin();
+        return el.type == type && (boost::stacktrace::hash_value(fl) ==
+                                   boost::stacktrace::hash_value(fr));
```

Comment:
is this duplicated below? looking at `deduplicate` and `contains`. if we do not want to use `operator==` we could also add a static compare function to `FpeInfo` and reference it here

Reply:
The problem is that the compare function is unary, so I'd have to make it a struct so it can capture the other operand. I can have the lambda call into a static comparator though.